ChemistryLib/PhreeqcInterface.cpp

```diff
@@ -0,0 +1,309 @@
+{
+    DBUG("Writing phreeqc inputs into file '%s'.", _phreeqc_input_file);
+    std::ofstream out(_phreeqc_input_file, std::ofstream::out);
+
+    if (!out)
+        ERR("Could not open file %s for writing phreeqc inputs.",
+            _phreeqc_input_file);
+
+    out << *this;
+
+    if (out.fail())
+        ERR("Failed in generating phreeqc input file %s.", _phreeqc_input_file);
+
+    out.close();
+}
+
+std::ofstream& operator<<(std::ofstream& out, phreeqcIO const& phreeqcIO)
+{
+    out << "SELECTED_OUTPUT" << "\n";
+    out << *phreeqcIO._output << "\n";
+
+    auto const& reaction_rates = phreeqcIO._reaction_rates;
+    if (!reaction_rates.empty())
+    {
+        out << "RATES" << "\n";
+        out << reaction_rates << "\n";
+    }
+
+    unsigned const num_nodes = phreeqcIO._solutions.size();
+    for (unsigned node_id = 0; node_id < num_nodes; ++node_id)
+    {
+        auto const& solution = phreeqcIO._solutions[node_id];
+        out << "SOLUTION " << node_id + 1 << "\n";
+        out << solution << "\n";
+
+        auto const& nodal_equilibrium_phases =
+            phreeqcIO._equilibrium_phases[node_id];
+        if (!nodal_equilibrium_phases.empty())
+        {
+            out << "EQUILIBRIUM_PHASES " << node_id + 1 << "\n";
+            out << nodal_equilibrium_phases << "\n";
+        }
+
+        auto const& nodal_kinetic_reactant =
+            phreeqcIO._kinetic_reactants[node_id];
+        if (!nodal_kinetic_reactant.empty())
+        {
+            out << "KINETICS " << node_id + 1 << "\n";
+            out << nodal_kinetic_reactant;
+            out << "-steps " << phreeqcIO._dt << "\n" << "\n";
+        }
+
+        out << "END" << "\n" << "\n";
+    }
+
+    return out;
+}
+
+void phreeqcIO::execute()
+{
+    auto database_loaded = [&](int instance_id) {
+        return LoadDatabase(instance_id, _database.c_str()) == 0;
+    };
+
+    DBUG("Phreeqc: Executing chemical calculation.");
+    // initialize phreeqc configurations
+    auto const instance_id = CreateIPhreeqc();
+
+    // load a specific database in the working directory
+    if (!database_loaded(instance_id))
+        ERR("Failed to load the specified thermodynamic database file %s.",
+            _database.c_str());
+
+    SetSelectedOutputFileOn(instance_id, 1);
+
+    if (RunFile(instance_id, _phreeqc_input_file) > 0)
+    {
+        OutputErrorString(instance_id);
+        ERR("Failed in performing speciation calculation with the generated "
+            "phreeqc input file '%s'.",
+            _phreeqc_input_file);
+    }
+}
+
+void phreeqcIO::readOutputsFromFile()
+{
+    auto const& basic_output_setups = _output->basic_output_setups;
+    auto const& phreeqc_result_file = basic_output_setups.getOutputFile();
+    DBUG("Reading phreeqc results from file '%s'.", phreeqc_result_file);
+    std::ifstream in(phreeqc_result_file);
+
+    if (!in.good())
+        ERR("Could not open file %s.", phreeqc_result_file);
+
+    in >> *this;
+
+    if (!in.eof())
```

Comment:
The check for EOF is not the same as checking if the stream is `good()`. See reference table https://en.cppreference.com/w/cpp/io/basic_ios/good
I'd recommend to replace both tests with `if (!in)`. Only if you have a special error message in case of EOF, test for that too.
This also applies to lines 110-117 above

Reply:
Fixed.

Reply:
`if (!in) ` following the parsing of phreeqc result file returns true, thereby leading to the killing of the program. This is because null terminator remains in the ifstream `in`. It looks fine using `if (!in.eof)` here.

Reply:
Handling EOF explicitly is good. Nevertheless the failbit and badbit should be checked too.

Reply:
I had no idea how to cover a check to this part. Could you please kindly offer me a suggested change?

Reply:
Fixed.